### DISCUSSION TITLE
Can widgets persist state?

### DISCUSSION THREAD

**Discussion by ketozhang** [Textualize/textual#226] [Q&A]
So far, the examples on the README always take state data from something external (mouse over from some watcher, time from system clock, etc). 

If I wanted my widget to print 1, 2, 3, 4, etc in a loop by storing some state variable, can we do that?

My use case is to store a renderable in the widget without the need to re-instantiate on every call to `Widget::render()`. Without this feature, Rich stuff like spinners and progress bars are very limited.

**Comment by ThatXliner** [comment]
Since it's just Python classes, can't you just take advantage of the fact that we have instance attributes? This is OOP at the core...

Or use `Reactive`

> The Hover class is a custom widget which displays a panel containing the classic text "Hello World". The first line in the Hover class may seem a little mysterious at this point:
> 
> ```python
> mouse_over = Reactive(False)
> ```
> This adds a mouse_over attribute to your class which is a bool with a default of False. Adding attributes like this makes them reactive: any changes will result in the widget updating.

https://github.com/Textualize/textual#creating-widgets

**Reply by ketozhang** [reply]
Yes, as long as you're not overwriting any existing APIs, be careful about setting class attributes as mutable objects which are shared among instance, or extend `Widget::__init__`. 

I guess my question was more about if Textual has set an API for storing state (e.g., `setState()` in React).

**Comment by igorfil** [comment]
Any further development on this discussion?

I really like Textual and would like to use it in my projects, however without clear handling of state if feels very limited.

For example, I would like to use Textual in an application. It runs a thread that loads some data and saves it as a state. Next, I'd like Textual to take that state and render. What's the Textual's way to do that?
